go/pools/numbered.go

```diff
@@ -117,12 +137,12 @@ func (nu *Numbered) GetAll() (vals []interface{}) {
 
 // GetOutdated returns a list of resources that are older than age, and locks them.
 // It does not return any resources that are already locked.
-func (nu *Numbered) GetOutdated(age time.Duration, purpose string) (vals []interface{}) {
+func (nu *Numbered) GetOutdated(age time.Duration, strict bool, purpose string) (vals []interface{}) {
 	nu.mu.Lock()
 	defer nu.mu.Unlock()
 	now := time.Now()
 	for _, nw := range nu.resources {
-		if nw.inUse || !nw.enforceTimeout {
+		if !strict && (nw.inUse || !nw.enforceTimeout) {
```

Comment:
The role of this flag is not obvious. What does it do? Are you fixing a different bug?
Also, is there a way to simplify this expression? Negations combined with `and`, `or`, and parenthesis. I'm completely lost :).

Reply:
haha sorry. I covered this in a sentence in the description, but you're right i should make it clearer in the code.

> I also added a strict mode to the GetOutdated. I can revert this if there are objections. But it seemed like when the pool is closed it should force all transactions to be closed. That could be a wrong assumption though.

let me know if that makes sense. If it does, i'll clear up the code. But don't want to bother with that if it turns out i was wrong.

Reply:
Oh. So you override dba workload flag if we're closing? If so, let's honor it: wait for dba connections to finish their work even if we're closing. Keeps it simple.

There is a higher level failsafe where a vttablet will hard crash if it's hung for too long trying to shut down.

Reply:
ok clearly i conflated too many things in 1 line. My bad... Let me break it down:

let's take the new dba functionality away for a min.  The old logic looked like this:

```go
if nw.inUse {
  continue
}
```

This causes it to not return outdated transactions that are currently in use.  I think this makes sense in almost all cases. However, I noticed that when [closing the pool with pool.Close()](https://github.com/youtube/vitess/pull/3645/files#diff-471a70f3c8221e8a5bf6eb026aa354b3L131) we use the same GetOutdated function.

To me it seemed to make sense that if you are closing the pool you should also cancel any transaction -- they wont ever be able to be committed/rolled back, will they? Since the pool is no longer accepting calls.

At least that was my thought.  In trying to type this out, I re-read everything and it seems like the pool will still respond to Commit/Rollback/Get's for existing transactions. It just won't be able to create new ones with Begin.  So maybe it makes sense to revert the `strict` stuff.

Reply:
Seems like that's sort of what you said too -- so pushing the revert now.

Reply:
`inUse` means that it's in the middle of a statement execution. So, we can't touch it. If it's in the middle of a transaction, it would be false.

There is a previous step that waits for all transactions to complete. So, all this code is just a failsafe in case things slip past that point. In other words, no need to complicate it. So, we can get rid of `strict`. Should be quite safe.

Reply:
Sounds good, and done! Thanks